packages/session/index.js

```diff
@@ -160,17 +161,38 @@ function populateAuthedItemMiddleware(keystone) {
     }
     req.user = item;
     req.authedListKey = list.key;
+    req.audiences = session.audiences;
 
     next();
   };
 }
 
-function startAuthedSession(req, { item, list }) {
+function restrictAudienceMiddleware({ public, audiences }) {
+  return (req, res, next) => {
+    if (public) {
+      // If the session restriction is marked public, we let everything through.
+      next();
+    } else if (
+      req.audiences &&
+      audiences &&
+      req.audiences.some(audience => audiences.includes(audience))
+    ) {
+      // Otherwise, if one of the session audiences matches one of the restriction audiences, we let them through.
+      next();
+    } else {
+      // If the don't make it through, we simply respond with a 403 Permission Denied
+      res.status(403).send();
```

Comment:
Bikeshed: A 403 here can possibly leak information about an endpoint existing, etc. It would be good to fall through to a generic 404 page instead.